### PR TITLE
Implementation of the batch transanction view

### LINKED ISSUE #1058: Add Batch Viewer to tx view

### PR DESCRIPTION
**Pull request by henrypalacios**
# Summary

Closes #1058

Implement the visualization of a batch of transactions through a node graph.

[Cytoscape](https://js.cytoscape.org/) has been used as interaction library due to its maturity and large community. But we had to apply styles in a way outside the project format.

image

### To Test:
- Go to [mainnet/tx/0xc2eb04b549d68c770062591e2561c5b1534ea665c6179d33a779c52376c9161d](https://pr1061--gpui.review.gnosisdev.com/tx/0xc2eb04b549d68c770062591e2561c5b1534ea665c6179d33a779c52376c9161d)
- Go to [mainnet/tx/0x51757dd21436d695824ceb13d21ec3c3ef5e5e5454a847951b1ea91e1c89b0b7](https://pr1061--gpui.review.gnosisdev.com/tx/0x51757dd21436d695824ceb13d21ec3c3ef5e5e5454a847951b1ea91e1c89b0b7)

### CODE REVIEW

**Comment by alongoni** [comment]
We are working with @henrypalacios adding some tooltips to the arrows in order to have the amounts traded by the users (Traders). Also we need to wrap the address of the AMMs.

We are exploring another layout with @henrypalacios called: "cose" and it's look like this:
Default view:
image
Expected view:
image

It' ll require some testing to achieve the result.

**Comment by alfetopito** [comment]
Funny edge case https://pr1061--gpui.review.gnosisdev.com/tx/0x51757dd21436d695824ceb13d21ec3c3ef5e5e5454a847951b1ea91e1c89b0b7

Screen Shot 2022-03-17 at 15 30 21

**Comment by alfetopito** [comment]
Well this is a giant mess https://pr1061--gpui.review.gnosisdev.com/tx/0x172bddae0015331f4b357905ff7995389597a38294a00b4b13e90c1d70884785/ 😅 

For reviewers: open https://dune.xyz/queries/205786/384521, filter by `barn` and sort by `num trades`

**Comment by elena-zh** [comment]
Could we add some kind of an anchor into the URL to separate what Tx view we are using? 
I mean, if we see Batch viewer, add `#batch ` into the URL or so? 
So a user will be able to share this link and open exactly the viewer, not the list. Also, when a user updates the page, it will also show the viewer. 
https://github.com/gnosis/cowswap/pull/2577#issuecomment-1076262692

**Comment by henrypalacios** [comment]
> Could we add some kind of an anchor into the URL to separate what Tx view we are using?

@elena-zh , as I commented to @ramirotw  that requirement will be followed in another PR because we already have a considerable size in this one.

File: src/apps/explorer/components/TransanctionBatchGraph/styled.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+export const STYLESHEET: Stylesheet[] = [
+  {
+    selector: 'node[label]',
+    style: {
+      label: 'data(label)',
+      color: '#cfcfcf',
+      height: 50,
+      width: 50,
+      'background-color': '#22232d',
```
Comment (matextrem) [inline src/apps/explorer/components/TransanctionBatchGraph/styled.ts:15]: We could move those colors to the variable files in a section related to the `BatchGraph`

**Review by matextrem — CHANGES_REQUESTED** [review]
Great Job guys 🚀   Just small comments regarding styles

Thinking out loud, BatchViewer builder class would be a good candidate to move to the sdk

**Comment by elena-zh** [comment]
> Could we add some kind of an anchor into the URL to separate what Tx view we are using?
> I mean, if we see Batch viewer, add `#batch ` into the URL or so?
> So a user will be able to share this link and open exactly the viewer, not the list. Also, when a user updates the page, it will also show the viewer.
> https://watch.screencastify.com/v/V0tvqMqCZzgtUIRcsfnK

@henrypalacios , here is a separate task for this #1072

**Comment by alongoni** [comment]
Thanks for the feedback @anxolin !

> Amazing job, really! tada
> 
> Some quick feedback:
> 
> * Edges: Can we add the logo for the token if available?

We need to try a solution for that, natively Cytoscape doesn't support images in the edges. 
> * Nodes: can we make users to have funny cow icons :)

I've added this cow :) 
image

> * Nodes: can we make AMMs have their respective logo?

@henry and @ramirotw are analyzing it.

> * [nice to have] Network: We can add a logo of the network too
> 
> image
> 
> One important thing, which can be out of the scope, but i'd love we take it into account at some point is that the way to see the chart is hidden and hard to find.
> 
> image
> 
> We discussed the other day with Michel, and one idea is to show, for example in the header, a summary of the batch: i.e. Traded volume, number of trades, and liquidity used (maybe we can even show sow the % of liquidity as a pie chart showin the AMMs) Then below the pie chart, would be a “view details” to see the chart

It's true. Maybe I can work designing those components and access to the batch viewer.

In general, we think there are many things we can improve. Maybe we can handle in a second iteration. E.g:
- Add anchor to the BV #1072
- Add a grid selector. To represent the graph in other layout.  
- Add token and AMMs logo.
- Change Traders icon randomly (Cow).

**Comment by henrypalacios** [comment]
@elena-zh Regarding your previous message :point_up: 

> 1. I think it would be nice to write a network name from the capital letter

:heavy_check_mark: done!

> 2. As for a trader icon, from my perspective lights from the cow eyes look a bit strange to me. However, might be skipped if is looks nicely to everyone else.

Looks like a funny cow, indifferent here!

> 3. Can we make token labels larger a bit? This schema is opened in a large screen, but token labels are almost not readable

I will review it with @alongoni.

> 4. Do we really need a clicked state on arrows/icons?

What is your suggestion at this point? make desktop hover only? Or do you mean the shading of the element?

> 5. As for amounts, I think we should use the same rounding rules as we use for the table. I mean, if we show an amount rounded to 8 decimals in the table view (example 0.00025354 ), ...

:heavy_check_mark:  done!

**Comment by elena-zh** [comment]
Hey @henrypalacios !

> > As for a trader icon, from my perspective lights from the cow eyes look a bit strange to me. However, might be skipped if is looks nicely to everyone else.
> 
> Looks like a funny cow, indifferent here!
 I do not insist on this as it was my personal impression! =)

> > Do we really need a clicked state on arrows/icons?
> 
> What is your suggestion at this point? make desktop hover only? Or do you mean the shading of the element?
Could we remove these on-click effects at all? Frankly, I can't understand why we need them.

Also, I can't test your changes due to failed build.. =((

Btw, what a nice view for the transaction https://pr1061--gpui.review.gnosisdev.com/tx/0x172bddae0015331f4b357905ff7995389597a38294a00b4b13e90c1d70884785/
image

Can we align diagram horizontally if something like this happens?

**Comment by elena-zh** [comment]
@henrypalacios , I have verified the latest changes!
Network names, rounding, and increased token names LGTM!
Maybe another nitpick I could mention is that it would be better to fill in the whole screen with the diagram area
image

**Comment by alongoni** [comment]
> @henrypalacios , I have verified the latest changes! Network names, rounding, and increased token names LGTM! Maybe another nitpick I could mention is that it would be better to fill in the whole screen with the diagram area image


Hey @elena-zh thanks! Is your zoom at 100%? Maybe that could be the cause.
https://www.loom.com/share/4c376579ba204d17a32fd3f02d4de014

**Comment by elena-zh** [comment]
Hey @alongoni , I have fund the root cause of the issue: the diagram area is not auto-scalable. I mean, if I drag and drop a browser's tab with the diagram from a  smaller screen to a bigger one, I will get this issue. See the video: https://watch.screencastify.com/v/u2oEJODe2166zN7faBdz

But it is a low-low priority issue, that is why I created a separate task for this #1084

**Comment by henrypalacios** [comment]
Hey @elena-zh, @ramirotw as I mentioned I submitted changes to mitigate issues #1074(pt. 1), #1084 in this first iteration. 

So we will continue working on the corresponding issues on a next iteration.

**Review by ramirotw — APPROVED** [review]

**Pull request merged**